include/proteus/CoreLLVMDevice.hpp

```diff
@@ -229,6 +232,174 @@ inline void specializeIR(
+    NewGV->copyAttributesFrom(GV);
+    VMap[GV] = NewGV;
+  }
+
+  for (auto *F : ReachableFunctions) {
+    auto *NewFunction = Function::Create(F->getFunctionType(), F->getLinkage(),
+                                         F->getAddressSpace(), F->getName(),
+                                         KernelModule.get());
+    NewFunction->copyAttributesFrom(F);
+    VMap[F] = NewFunction;
+  }
+
+  for (auto *F : ReachableDeclarations) {
+    auto *NewFunction = Function::Create(F->getFunctionType(), F->getLinkage(),
+                                         F->getAddressSpace(), F->getName(),
+                                         KernelModule.get());
+    NewFunction->copyAttributesFrom(F);
+    NewFunction->setLinkage(GlobalValue::ExternalLinkage);
+    VMap[F] = NewFunction;
+  }
+
+  for (GlobalVariable *GV : ReachableGlobals) {
+    GlobalVariable *NewGV = cast<GlobalVariable>(VMap[GV]);
+    NewGV->setInitializer(MapValue(GV->getInitializer(), VMap));
+  }
+
+  for (auto *F : ReachableFunctions) {
+    SmallVector<ReturnInst *, 8> Returns;
+    auto *NewFunction = dyn_cast<Function>(VMap[F]);
+    Function::arg_iterator DestI = NewFunction->arg_begin();
+    for (const Argument &I : F->args())
+      if (VMap.count(&I) == 0) {
+        DestI->setName(I.getName());
+        VMap[&I] = &*DestI++;
+      }
+    llvm::CloneFunctionInto(NewFunction, F, VMap,
+                            CloneFunctionChangeType::DifferentModule, Returns);
+  }
+
+  // Copy annotations from M into KernelModule now that VMap has been populated.
+  const std::string MetadataToCopy[] = {"llvm.annotations", "nvvm.annotations",
+                                        "nvvmir.version", "llvm.module.flags"};
+  for (auto &MetadataName : MetadataToCopy) {
+    NamedMDNode *NamedMD = M.getNamedMetadata(MetadataName);
+    if (!NamedMD)
+      continue;
+
+    auto *NewNamedMD = KernelModule->getOrInsertNamedMetadata(MetadataName);
+    for (unsigned I = 0, E = NamedMD->getNumOperands(); I < E; ++I) {
+      MDNode *MDEntry = NamedMD->getOperand(I);
+      bool ShouldClone = true;
+      // Skip if the operands of an MDNode refer to non-existing,
+      // unreachable global values.
+      for (auto &Operand : MDEntry->operands()) {
+        Metadata *MD = Operand.get();
+        auto *CMD = dyn_cast<ConstantAsMetadata>(MD);
+        if (!CMD)
+          continue;
+
+        auto *GV = dyn_cast<GlobalValue>(CMD->getValue());
+        if (!GV)
+          continue;
+
+        if (!VMap.count(GV)) {
+          ShouldClone = false;
+          break;
+        }
+      }
+
+      if (!ShouldClone)
+        continue;
+
+      NewNamedMD->addOperand(MapMetadata(MDEntry, VMap));
+    }
+  }
+
+#if PROTEUS_ENABLE_DEBUG
+  if (verifyModule(*KernelModule, &errs()))
+    PROTEUS_FATAL_ERROR("Broken mini-module found, JIT compilation aborted!");
+
+  Logger::logs("proteus") << "Mini-module \n" << *KernelModule << "\n";
```

Comment:
Probably want to put the log above the verify - since if we get a fatal error we will never see the module.